morph_tool/morphdb.py

```diff
@@ -0,0 +1,392 @@
+            obj.df[key] = True
+            obj.df['axon_inputs'] = [[] for _ in range(len(obj.df))]
+        return obj
+
+    @classmethod
+    def _from_neurondb_xml(cls, neurondb, morph_paths, label):
+        obj = MorphDB()
+        with neurondb.open() as fd:
+            content = fd.read()
+            neurondb = xmltodict.parse(content)
+
+        morphologies = neurondb['neurondb']['listing']['morphology']
+
+        # Case where there is a single <morphology></morphology> tag in the XML
+        if not isinstance(morphologies, list):
+            morphologies = [morphologies]
+        morphologies = filter(None, morphologies)
+        morphologies = list(map(MorphInfo._from_xmldict, morphologies))  # noqa, pylint: disable=protected-access
+
+        for morph in morphologies:
+            morph.label = label
+            morph.path = morph_paths.get(morph.name)
+
+        obj.df = MorphDB._create_dataframe(morphologies)
+        obj.df = obj.df.astype({key: bool for key in BOOLEAN_REPAIR_ATTRS})
+        return obj
+
+    @classmethod
+    def from_neurondb(cls,
+                      neurondb: Path,
+                      label: str = 'default',
+                      morphology_folder: Optional[Path] = None):
+        '''Builds a MorphologyDB from a neurondb.(xml|dat) file
+        Args:
+            neurondb: path to a neurondb.(xml|dat) file
+            label: a unique label to mark all morphologies coming from this neurondb
+            morphology_folder: the location of the morphology files, if None it will default
+                to the neurondb folder
+
+        Raises: ValueError if the neurondb does not abide by the specification
+        https://bbpteam.epfl.ch/documentation/projects/morphology-repair-workflow/latest/input_files.html#specification
+
+        ..note:: missing keys are filled with `True` values
+        '''
+        morphology_folder = morphology_folder or neurondb.parent.resolve()
+
+        morph_paths = {path.stem: path for path in iter_morphology_files(morphology_folder)}
+
+        if neurondb.suffix.lower() == '.dat':
+            return cls._from_neurondb_dat(neurondb, morph_paths, label)
+        else:
+            return cls._from_neurondb_xml(neurondb, morph_paths, label)
+
+    @classmethod
+    def from_folder(cls,
+                    morphology_folder: Path,
+                    mtypes: Iterable[Tuple[str, str]],
+                    label: str = 'default'):
+        '''Factory method to create a MorphDB object from a folder containing morphologies
+
+        Args:
+            morphology_folder: a folder containing morphologies
+            mtypes: a sequence of 2-tuples (morphology name, mtype)
+            label: (optional) a group label to be used to identify the morphlogies from this folder
+
+        Raises: ValueError if the folder contains multiple files with the same name but
+        different extensions
+        '''
+        files = list(iter_morphology_files(morphology_folder))
```

Comment:
I'm wondering if it could be more efficient to filter `files` based on the morphology names passed in `mtypes` (converted to a dict or set). It depends on the number of files, and the number of mtypes passed to the function.
However if there could be different cases I think it's fine as is, and we can improve it later if needed.

Reply:
The general use case is that `iter_morphology_files` and `mtypes` should return the same files and should be the same length so that should not make a big difference I think.